# Filtering accessible sites

Attempts to filter out accessible sites, following Alastairs selecting variants and samples blog post.

In [11]:
##Import modules and data
import h5py
import numpy as np
import sys
import allel

In [2]:
callset = h5py.File("../data/phase1/hdf5/ag1000g.phase1.ar3.pass.h5", mode='r')
chrom = '3L'

Not sure whether to shallow copy or not, so here are both

In [4]:
accessible = callset[chrom]['variants']['Accessible'][:]
accessible_nocolon = callset[chrom]['variants']['Accessible']

In [5]:
accessible

array([ True,  True,  True, ...,  True,  True,  True])

In [6]:
accessible_nocolon

<HDF5 dataset "Accessible": shape (9643193,), type "|b1">

accessible here appears to be an array of boolean values, not sure what the _nocolon is


In [7]:
##check info about variables
%whos


Variable             Type       Data/Info
-----------------------------------------
accessible           ndarray    9643193: 9643193 elems, type `bool`, 9643193 bytes (9.196465492248535 Mb)
accessible_nocolon   Dataset    <HDF5 dataset "Accessible<...>e (9643193,), type "|b1">
callset              File       <HDF5 file "ag1000g.phase1.ar3.pass.h5" (mode r)>
chrom                str        3L
h5py                 module     <module 'h5py' from '/hom<...>ckages/h5py/__init__.py'>
np                   module     <module 'numpy' from '/ho<...>kages/numpy/__init__.py'>
sys                  module     <module 'sys' (built-in)>


False counts as 0 in python and True 1, so if only some of the data are true the value we'll get will be less than the total number of values, 9643193

In [8]:
#count all of the non-zero values
np.count_nonzero(accessible)

9643193

In [9]:
##try adding all values in array together
sum(accessible)

9643193

In [10]:
##check to see what values found in array
np.unique(accessible_nocolon)

array([ True])

These all seem to suggest that all of the sites in the pass data are accessible.

Now to try in a non-filtered file that we know has both accessible and non accessible sites, from the accessibility data folder.

First we need to convert to hdf5 because I don't know how to work in vcf:

In [16]:
##convert using scikit-allel
allel.vcf_to_hdf5('../data/phase1/vcf/accessibility.3L.vcf', '../data/phase1/vcf/accessibility.3L.h5', fields='*', overwrite=True)

Now to load the file into our environment.

In [19]:
test = h5py.File('../data/phase1/vcf/accessibility.3L.h5', mode='r')
test

<HDF5 file "accessibility.3L.h5" (mode r)>

Here we'll set up two variables, one that is shallow copied and one that isn't (this will act as a count for total number of sites.

In [23]:
test_access = test['variants']['Accessible'][:]
test_access

array([False, False, False, ..., False, False, False])

In [25]:
test_numbers = test['variants']['Accessible']
test_numbers

<HDF5 dataset "Accessible": shape (41963435,), type "|b1">

Now lets sum the shallow array:

In [24]:
sum(test_access)

24946488

If we sum the shallow copied one, we see that it =/= the total count. This therefore tells us that all sites in the pass data, at least for chromosome 3L, are accessible.